### DISCUSSION TITLE
Wrong behavior handling path parameter encoding

### DISCUSSION THREAD

**Discussion by vmaurin** [fastapi/fastapi#11369] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()

@app.get("/example/{object_type}/{object_id}")
def show_example(object_type: str, object_id: str):
    return {"params": [object_type, object_id]}
```

### Description

In this kind of scenario, everything is fine as long as the path parameters don't contain any special chars. When they do, the usual way of doing it, is to URL encode them, but it seems in FastAPI it doesn't work (the decoding looks to happen before the routing logic)

I don't see any good reason of doing it

Here few uses cases

* `curl "http://localhost:8000/example/post/123"` => current and expected behavior: `{"params":["post","123"]}` :+1: 
* `curl "http://localhost:8000/example/post/foo/bar"` => current and expected behavior: return a 404 :+1: 
* `curl "http://localhost:8000/example/post/foo%2Fbar"` => current behavior : 404, expected behavior : `{"params":["post","foo/bar"]}` :-1: 
* `curl "http://localhost:8000/example/post/foo%252Fbar"` => current and expected behavior: `{"params":["post","foo%2Fbar"]}` :+1: 

There is the `:path` prefix of Starlette that could help a bit, but it doesn't solve the issue if both parameters could contain special encoded chars

### Operating System

Linux

### FastAPI Version

0.105.0

### Pydantic Version

2.6.3

### Python Version

3.11

**Comment by jgould22** [comment, marked as answer]
This has come up before 

https://github.com/tiangolo/fastapi/issues/791

and 

https://github.com/encode/starlette/issues/826

**Reply by vmaurin** [reply]
Thank you ! I didn't found it still doing search (like "path parameter encoding")
